### PR TITLE
Introduce `EntityManagerInterface#wrapInTransaction()`

### PR DESCRIPTION
**Pull request by simPod**
This introduces `Doctrine\ORM\EntityManagerInterface#wrapInTransaction()` so we can return any value from Transaction.

### CODE REVIEW

**Comment by greg0ire** [comment]
> I believe I've seen this bug fixed on v3 like half a year ago. Now I've found this PR stale so I just wanted to update the test and I've found the fix is not there anymore. I can't tell what happened here, branches are shuffled etc.

The v3 you know has been renamed to old-prototype-3.x, and we are working on backporting some of the PRs here: https://github.com/doctrine/orm/projects/4

The difference this time is we are going to regularly merge up 2.x into 3.x .

We may backport more, feel free to find the commits in old-prototype-3.x and backport them 👍

**Comment by greg0ire** [comment]
Is there a BC-break in this PR? If not, please target 2.10.x instead.

**Comment by greg0ire** [comment]
Or even 2.9.x since it is a bugfix

**Comment by simPod** [comment]
Thanks for explaining, I'll revert this and target the`old-prototype`. This was initially only about tests for the bug.

**Comment by simPod** [comment]
@greg0ire I suppose CI is broken on old-prototype, right?

**Comment by greg0ire** [comment]
Yes, it's no longer maintained, it's just a source of commits to port to maintained branches. Why are you targeting it? Does the bug only happen there?

**Comment by simPod** [comment]
@greg0ire the bug is fixed only there 

https://github.com/doctrine/orm/commit/aa9f11e11d31590c6c346074f33ea7bf01468fd1#diff-09261ffd6eb75f4fda9de6c2fbf156761380cf13687e57a199878630890402c4R232

the previous version (e.g. the whole 2.x orm) returns `true` from `transactional()` when inner closure returns anything falsy (`[]`, `0`, `false`) instead of the falsy value.

**Comment by greg0ire** [comment]
Could you please work on porting the new behavior to 3.x instead? Otherwise it's just more porting work for us in the future

**Comment by simPod** [comment]
ah ok, I have just reverted that few h ago. Np. Gimme sec.

**Comment by simPod** [comment]
It's done, test failures are irrelevant https://github.com/simPod/orm/runs/3093712747.

**Comment by greg0ire** [comment]
I think they relate to https://github.com/doctrine/cache/pull/381

**Comment by greg0ire** [comment]
Blocked by https://github.com/doctrine/orm/pull/8847 and 2 merges up

File: lib/Doctrine/ORM/EntityManager.php [hunk 1]
```diff
@@ -237,7 +237,9 @@ public function transactional($func)
             $this->flush();
             $this->conn->commit();
 
-            return $return ?: true;
+            // Can't use ?? because PHP 7.1 is supported
+            //phpcs:ignore SlevomatCodingStandard.ControlStructures.RequireNullCoalesceOperator
+            return $return === null ? true : $return;
```
Comment (greg0ire) [inline lib/Doctrine/ORM/EntityManager.php]: This description looks wrong now:

https://github.com/doctrine/orm/blob/551018136bc0c4f53f047a33f7fdac128589c6b7/lib/Doctrine/ORM/EntityManagerInterface.php#L83

Also, BC-breaks should be documented in UPGRADE.md

**Comment by greg0ire** [comment]
Regarding the PR itself: why did non-empty value deserve a special treatment, and why does `null` deserve one?

**Comment by simPod** [comment]
@greg0ire I misinterpreted `non-empty`. Will fix.

File: lib/Doctrine/ORM/EntityManager.php (hunk 1 above)
Comment (simPod) [reply lib/Doctrine/ORM/EntityManager.php]: done

**Comment by greg0ire** [comment]
Please fix the unit tests. The failures are now relevant.

**Comment by beberlei** [comment]
This is a problematic BC break because we cannot offer a forward compatible way to migrate or a way to detect the deprecated behavior and warn about it.

I believe this requires a new method name.

**Comment by simPod** [comment]
Hmm, would name should we choose? Do you have some Doctrine guidelines on method names? `transact()` / `transaction()` / `?`?

**Comment by greg0ire** [comment]
How about `wrapInTransaction()`?

**Comment by simPod** [comment]
Ok, initially I just wanted to add few more tests cases to ensure it returns values properly. This is gonna be slightly bigger tho 😄

**Comment by greg0ire** [comment]
Also, since this should be backwards compatible, you should target 2.10.x

**Comment by simPod** [comment]
So, should I introduce a new method, that returns all value types and
target 2.10.x? Am I right?

**Comment by greg0ire** [comment]
Yes! It's better because:
- we have an upgrade path
- we can try this out earlier, and know if it's a good idea without having to wait for 3.0.0

**Comment by simPod** [comment]
#8872

File: lib/Doctrine/ORM/EntityManager.php [hunk 2]
```diff
@@ -248,6 +248,28 @@ public function transactional($func)
         }
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public function wrapInTransaction(callable $func)
```
Comment (greg0ire) [inline lib/Doctrine/ORM/EntityManager.php:254]: Shouldn't `transactional()` be deprecated in favor of this method?

Comment (simPod) [reply lib/Doctrine/ORM/EntityManager.php:254]: Aight, tried to deprecate it :)

File: lib/Doctrine/ORM/Decorator/EntityManagerDecorator.php [hunk 3]
```diff
@@ -53,6 +53,14 @@ public function transactional($func)
         return $this->wrapped->transactional($func);
     }
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Decorator/EntityManagerDecorator.php:61]: This method probably needs to be deprecated too

Comment (simPod) [reply lib/Doctrine/ORM/Decorator/EntityManagerDecorator.php:61]: It implements the interface so should be covered by that 🤔

**Comment by greg0ire** [comment]
Please fix the conflict :pray: 

Note: this PR kind of backports https://github.com/doctrine/orm/pull/6147

**Comment by simPod** [comment]
done

> Note: this PR kind of backports #6147

I believe I've seen it multiple times here so initially I only wanted to add test cases for it then I§ve found out it's not implemented 🤷🏾

**Comment by greg0ire** [comment]
Hence the branch name!

**Review by greg0ire — DISMISSED** [review]

**Comment by SenseException** [comment]
The extended interface is a BC break. @greg0ire Is this considered as a "smaller" one?

File: lib/Doctrine/ORM/EntityManagerInterface.php [hunk 4]
```diff
@@ -80,12 +80,30 @@ public function beginTransaction();
...
+     * {@link flush} is invoked prior to transaction commit.
+     *
+     * If an exception occurs during execution of the function or flushing or transaction commit,
+     * the transaction is rolled back, the EntityManager closed and the exception re-thrown.
+     *
+     * @param callable $func The function to execute transactionally.
+     *
+     * @return mixed The value returned from the closure.
+     */
+    public function wrapInTransaction(callable $func);
```
Comment (greg0ire) [inline lib/Doctrine/ORM/EntityManagerInterface.php]: @SenseException maybe let's just do this:
```suggestion
    // public function wrapInTransaction(callable $func);
```

callers will have to check that this method exists before calling it though.

Comment (SenseException) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: This will do it for 2.10.

Comment (simPod) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: or I can target 3.0?

Comment (greg0ire) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: @simPod please don't retarget this whole PR to 3.0, that would mean we cannot remove `transactional` in 3.0

Comment (simPod) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: I'm waiting for instructions 🙃

Comment (greg0ire) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: Ok, here are the steps:

1. apply my suggestion
2. squash it with the previous commit
3. improve the upgrade notes by explaining how people are supposed to check that `wrapInTransaction` exists before calling it, if they have an instance of `EntityManagerInterface`

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @simPod !

File: lib/Doctrine/ORM/EntityManagerInterface.php (hunk 4 above)
Comment (zoltan-schwarz) [reply lib/Doctrine/ORM/EntityManagerInterface.php]: I'm really curious. What is the reason to comment this out? Why can't it be the interface?

**Pull request merged**